### PR TITLE
New distributions

### LINKED ISSUE #2268: MeixnerDistribution: slow pdfgradient
MeixnerDistribution doesnt implement pdf gradients, in turn the generic method is slow because changing parameters call the update methods which relies on optimization to update various coefficients

see t_Distribution_std.py test: 2/3 of the time is spent there

### PR DESCRIPTION
**Pull request by regislebrun**
This PR proposes these new features:
  * The UniformOverMesh class.  This class allows to define a uniform distribution over an arbitrary mesh. The sampling method relies on new capabilities in DistFunc related to uniform sampling in a simplex. As a result, the distFunc class has gained new methods to uniformly sample segments, triangles, tetrahedrons and simplexes.
  * The PiecewiseLinearDistribution. This class allows to define a continuous distribution by a piecewise linear function describing its density. It will be restricted to the univariate case in a first step.
  * The extension of the DiGamma function to complex arguments, in order to fix #2268 
  * A specific implementation of the moments computation for the Rice distribution, in order to speed-up these computations and gain extra digits. It is linked to #2226
  * A rework of the PiecewiseLinearEvaluation and PiecewiseHermiteEvaluarion classes

### CODE REVIEW

**Comment by jschueller** [comment]
remember new classes must go to the experimental module

File: python/src/dist_bundle3_module.i
Comment (jschueller) [inline python/src/dist_bundle3_module.i:50]: remember new classes must go to the experimental module

Comment (regislebrun) [reply python/src/dist_bundle3_module.i]: It was what I had in mind, but forgot it at the last minute :-(

File: lib/src/Uncertainty/Distribution/MeixnerDistribution.cxx
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/MeixnerDistribution.cxx:419]: no need to overload description methods and str/repr (since the function is not exposed to user)
only clone/getIn/Outputdimension are strictly needed

File: lib/src/Base/Func/SpecFunc/openturns/SpecFunc.hxx
Comment (jschueller) [inline lib/src/Base/Func/SpecFunc/openturns/SpecFunc.hxx:181]: this can be probably saved for another PR

File: python/test/t_UniformOverMesh_std.py
Comment (jschueller) [inline python/test/t_UniformOverMesh_std.py:82]: expensive

**Comment by jschueller** [comment]
wait, the piecewiselienardistribution is not finished, why did you mark it as ready ?

File: lib/src/Uncertainty/Distribution/PiecewiseLinearDistribution.cxx
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/PiecewiseLinearDistribution.cxx:104]: weird, thats the uniform code

**Comment by jschueller** [comment]
I got it: the piecewisedistribution is not enabled in cmakelists

File: lib/src/Uncertainty/Distribution/PiecewiseLinearDistribution.cxx
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/PiecewiseLinearDistribution.cxx:104]: @jschueller I am currently removing the PiecewiseLinearDistribution class from this PR as it will not be ready for the RC. The other classes are ready for a review.

Comment (jschueller) [reply lib/src/Uncertainty/Distribution/PiecewiseLinearDistribution.cxx:104]: got it

File: lib/src/Uncertainty/Distribution/MeixnerDistribution.cxx
Comment (jschueller) [reply lib/src/Uncertainty/Distribution/MeixnerDistribution.cxx:419]: also they should be marked override

File: lib/src/Uncertainty/Distribution/openturns/DistFunc.hxx
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/openturns/DistFunc.hxx:317]: the new SpecFunc, DistFunc methods must be listed in the usermanual

**Comment by sofianehaddad** [comment]
Only minor comments (+ remarks from Julien)

File: lib/src/Uncertainty/Distribution/MeixnerDistribution.cxx
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/MeixnerDistribution.cxx:419]: These methods were here for debug purpose. I will remove them.

File: lib/src/Uncertainty/Distribution/DistFunc.cxx
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx:1980]: There is no actual advantage in generating the pseudo-random numbers in advance rather than on the fly if:
 * you don't apply an expensive transformation on these numbers
 * you don't apply it in parallel

The way modern pseudo-random numbers generators work is to manage a huge state, slice bits in it to generate unitary realizations then perform a big update once the current state has been exhausted.

File: lib/src/Uncertainty/Distribution/openturns/DistFunc.hxx
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/openturns/DistFunc.hxx:317]: @jschueller The documentation in commit ff7f62c136c4c57137cff0cceedd3183dc9eb0a6 does not fit your expectations?

Comment (jschueller) [reply lib/src/Uncertainty/Distribution/openturns/DistFunc.hxx:317]: mostly :)
but the SpecFunc functions must be explicitely listed:
https://github.com/openturns/openturns/blob/master/python/doc/user_manual/functions.rst
same for DistFunc methods in probabilistic_modelling.rst

File: lib/src/Uncertainty/Distribution/DistFunc.cxx
Comment (jschueller) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx:1968]: you're probably more used to see "a + u * (b-a)", but that ends up in the same result

Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx:1968]: I wrote things this way deliberately, as the key in sampling uniformly a simplex is to sample uniformly the barycentric coordinates in the standard simplex and return the associated convex combinations of the simplex vertices. Of course the coordinates have to sum to 1, hence the (u, 1-u) pair. In 2d you get (u, v, 1-u-v) but the benchmark says that it is more efficient to select a segment parallel to an edge then to sample uniformly this segment. In higher dimensions the barycentric method is the best.

File: lib/src/Uncertainty/Distribution/openturns/DistFunc.hxx
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/openturns/DistFunc.hxx:317]: I got it! For SpecFunc, it was just an extension of existing functions to the complex domain so they were already in the User Manual. I just added the new functions proposed by DistFunc.

File: python/src/UniformOverMesh_doc.i.in
Comment (mbaudin47) [inline python/src/UniformOverMesh_doc.i.in:77]: The default value is?

File: python/src/UniformOverMesh_doc.i.in
Comment (mbaudin47) [inline python/src/UniformOverMesh_doc.i.in:57]: The default value is?

File: python/src/UniformOverMesh_doc.i.in
Comment (mbaudin47) [inline python/src/UniformOverMesh_doc.i.in:37]: There are two specific methods here: `getMesh` and `getIntegrationAlgorithm`. But the help page has more than 50 methods, because of class inheritance. I would suggest to add 1 or 2 example lines in the main example to show this, otherwise a user may not see that point.

File: lib/src/Uncertainty/Distribution/UniformOverMesh.cxx
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/UniformOverMesh.cxx:163]: I can vaguely see why that attribute is there, but it is not used by the algorithm within the class and there is no accessor: is is a temporary debug code ? shall it be removed?

File: lib/src/Uncertainty/Distribution/UniformOverMesh.cxx
Comment (mbaudin47) [inline lib/src/Uncertainty/Distribution/UniformOverMesh.cxx:130]: Are you certain that `i` is a correct naming here?

**Review by mbaudin47 — CHANGES_REQUESTED** [review]
The new `UniformOverMesh` class is graphically spectacular! I only have a few comments and questions.

File: python/src/UniformOverMesh_doc.i.in
Comment (regislebrun) [reply python/src/UniformOverMesh_doc.i.in:57]: Suspens... `IteratedQuadrature` ;-)! From time to time I am called by someone right in the middle of a sentence...

File: lib/src/Uncertainty/Distribution/UniformOverMesh.cxx
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/UniformOverMesh.cxx:163]: Good catch! It was used to sample the distribution, but the service has been optimized and moved to the DistFunc class.

File: lib/src/Uncertainty/Distribution/UniformOverMesh.cxx
Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/UniformOverMesh.cxx:130]: i like interval ;-) Ok I will change it

**Review by jschueller — APPROVED** [review]
stunning!
too bad there is no dedicated gallery example in the doc (the 2-d sin example is great)

**Comment by jschueller** [comment]
yes, waiting for sofiane review
did you fix the default value remark of michael ?

File: python/src/UniformOverMesh_doc.i.in
Comment (regislebrun) [reply python/src/UniformOverMesh_doc.i.in:83]: Oupss I fixed the first one, not this one. Sorry.

**Comment by regislebrun** [comment]
@sofianehaddad I wrote a benchmark to check how the different algorithms compare between them, and how a call to rUniformSegment, rUniformTriangle and rUniformTetrahedron from within rUniformSimplex compare to a direct call to these methods. I did it both at the C++ level and the Python level. The results are:
 * the generic (i.e. valid in any dimension) algorithm implemented in rUniformSimplex is roughly 10x slower than the one in rSegment, 5x slower than the one in rTriangle and 2x slower than the one in rTetraedron.
 * Among the many options to sample uniformly in a triangle, the algorithm based on a square root is the most efficient (at least by 10%) , which I still don't understand.
 * There is *no* penalty involved by the creation of temporary points when calling rUniformSegment and so on from rUniformSimplex compared to a direct call to these methods. I first implemented the selection of methods in UniformOverMesh::getRealization() but after the benchmark I decided to push it into DistFunc::rUniformSimplex()

File: lib/src/Uncertainty/Distribution/DistFunc.cxx
Comment (sofianehaddad) [inline lib/src/Uncertainty/Distribution/DistFunc.cxx:2142]: Do you mean `result(n, 0) = vertices(0, 0)`?

File: lib/src/Uncertainty/Distribution/DistFunc.cxx
Comment (sofianehaddad) [inline lib/src/Uncertainty/Distribution/DistFunc.cxx:2146]: Did not understand ?
`result(n,j) += ...` instead or `=`?

**Comment by sofianehaddad** [comment]
Understand your benchmark' results accurate with thé expectation.
I am more focusing on `UniformOverMesh::getSample`

**Comment by jschueller** [comment]
A rebase is needed after the chaos merge

File: lib/src/Uncertainty/Distribution/DistFunc.cxx
Comment (jschueller) [inline lib/src/Uncertainty/Distribution/DistFunc.cxx:2142]: I dont understand this last change as the loop only alters one memory address, should it be result(n,j) ?

Comment (jschueller) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx:2142]: If you're going to add another change, you can mention the new class in the changelog while you're at it

File: python/doc/user_manual/probabilistic_modelling.rst
Comment (jschueller) [inline python/doc/user_manual/probabilistic_modelling.rst:307]: rUniformTetrahedron also

File: lib/src/Uncertainty/Distribution/DistFunc.cxx
Comment (sofianehaddad) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx:2142]: indeed :-(

**Comment by regislebrun** [comment]
I implemented a UniformOverMesh::getSample() method, with no speed improvement but additional memory consumption. The code was the following, trying to benefit from a potential speed-up calling simplexSelection_.getSample() instead of simplexSelection_.getRealization() within a loop.
```
Sample UniformOverMesh::getSample(const UnsignedInteger size) const
{
  const Sample vertices(mesh_.getVertices());
  const Sample simplicesIndices(simplexSelection_.getSample(size));
  const UnsignedInteger dimension = getDimension();
  Sample result(size, dimension);
  for (UnsignedInteger i = 0; i < size; ++i)
    {
      const Indices simplex(mesh_.getSimplex(static_cast<UnsignedInteger>(simplicesIndices(i, 0))));
      const Point x(DistFunc::rUniformSimplex(vertices.select(simplex)));
      for (UnsignedInteger j = 0; j < dimension; ++j)
        result(i, j) = x[j];
    } // i
  return result;
}
```

**Comment by regislebrun** [comment]
I finally switched from the IteratedQuadrature algorithm to a GaussLegendre algorithm as the default integration algorithm for the computation of the CDF. It is controlled by two keys in ResourceMap, see the documentation.

File: lib/src/Uncertainty/Distribution/DistFunc.cxx
Comment (sofianehaddad) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx:1968]: Basic question : using `u * a + (1.0 - u) * b` I guess we temporarly create `Point` objects (`u * a`, `(1-u)*b`), isn't it?

Comment (regislebrun) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx:1968]: Maybe. As usual, there is a trade-off between performance and readability.  The best is to set-up a benchmark and see if there is something to gain here. Using objdump I was not able to see if gcc had made some magic here. I just got the flamegraph, I will check it tomorrow.
Is it reasonable to postpone the benchmarks to the RC phase? I have additional code to write upon this PR...

**Comment by sofianehaddad** [comment]
> I implemented a UniformOverMesh::getSample() method, with no speed improvement but additional memory consumption. The code was the following, trying to benefit from a potential speed-up calling simplexSelection_.getSample() instead of simplexSelection_.getRealization() within a loop.
> 
> ```
> Sample UniformOverMesh::getSample(const UnsignedInteger size) const
> {
>   const Sample vertices(mesh_.getVertices());
>   const Sample simplicesIndices(simplexSelection_.getSample(size));
>   const UnsignedInteger dimension = getDimension();
>   Sample result(size, dimension);
>   for (UnsignedInteger i = 0; i < size; ++i)
>     {
>       const Indices simplex(mesh_.getSimplex(static_cast<UnsignedInteger>(simplicesIndices(i, 0))));
>       const Point x(DistFunc::rUniformSimplex(vertices.select(simplex)));
>       for (UnsignedInteger j = 0; j < dimension; ++j)
>         result(i, j) = x[j];
>     } // i
>   return result;
> }
> ```

How is it possible to have additional memory consumption?  There is no **additional** things in comparison to `ContinuousDistribution::getSample`

I had less generic implementation when we intentionnaly redirect method depending on the dimension. Something like : 

```cxx
Sample UniformOverMesh::getSample(const UnsignedInteger size) const
{
  const Sample vertices(mesh_.getVertices());
  const UnsignedInteger dimension = vertices.getDimension();
  const Point simplicesIndices(simplexSelection_.getSample(size).asPoint());
  Sample result(size, dimension);

  if (dimension == 1)
  {
    for (UnsignedInteger i = 0; i < size; ++i)
    {
      const Indices simplex(mesh_.getSimplex(static_cast<UnsignedInteger>(simplicesIndices[i])));
      for (UnsignedInteger j = 0; j < dimension; ++j)
        result(i, j) = vertices(simplex[0], j);
    } // i
    return result;
  }

  if (dimension == 2)
  {
    Point a(dimension), b(dimension);
    Pointt x(dimension);
    for (UnsignedInteger i = 0; i < size; ++i)
    {
      const Indices simplex(mesh_.getSimplex(static_cast<UnsignedInteger>(simplicesIndices[i])));
      // Indices[0] ==> a
      // Indices[1] ==> b
      for (UnsignedInteger j = 0; j < dimension; ++j)
      {
        a[j] = vertices(simplex[0]), j);
        b[j] = vertices(simplex[1]), j);
      }
      x = DistFunc::rUniformSegment(a, b);
      for (UnsignedInteger j = 0; j < dimension; ++j)
        result(i, j) = x[j];
    } // i
    return result;
  }

  if (dimension == 3)
  {
    Point a(dimension), b(dimension), c(dimension);
    Pointt x(dimension);
    for (UnsignedInteger i = 0; i < size; ++i)
    {
      const Indices simplex(mesh_.getSimplex(static_cast<UnsignedInteger>(simplicesIndices[i])));
      // Indices[0] ==> a
      // Indices[1] ==> b
      // Indices[2] ==> c
      for (UnsignedInteger j = 0; j < dimension; ++j)
      {
        a[j] = vertices(simplex[0]), j);
        b[j] = vertices(simplex[1]), j);
        c[j] = vertices(simplex[2]), j);
      }
      x = DistFunc::rUniformTriangle(a, b, c);
      for (UnsignedInteger j = 0; j < dimension; ++j)
        result(i, j) = x[j];
    } // i
    return result;
  }

  if (dimension == 4)
  {
    Point a(dimension), b(dimension), c(dimension), d(dimension);
    Pointt x(dimension);
    for (UnsignedInteger i = 0; i < size; ++i)
    {
      const Indices simplex(mesh_.getSimplex(static_cast<UnsignedInteger>(simplicesIndices[i])));
      // Indices[0] ==> a
      // Indices[1] ==> b
      // Indices[2] ==> c
      // Indices[3] ==> d
      for (UnsignedInteger j = 0; j < dimension; ++j)
      {
        a[j] = vertices(simplex[0]), j);
        b[j] = vertices(simplex[1]), j);
        c[j] = vertices(simplex[2]), j);
        d[j] = vertices(simplex[3]), j);
      }
      x = DistFunc::rUniformTetrahedron(a, b, c, d);
      for (UnsignedInteger j = 0; j < dimension; ++j)
        result(i, j) = x[j];
    } // i
    return result;
  }

  // other case
  for (UnsignedInteger i = 0; i < size; ++i)
    {
      const Indices simplex(mesh_.getSimplex(static_cast<UnsignedInteger>(simplicesIndices[i])));
      const Point x(DistFunc::rUniformSimplex(vertices.select(simplex)));
      for (UnsignedInteger j = 0; j < dimension; ++j)
        result(i, j) = x[j];
    } // i
  return result;
}
```
By the way, I think we could merge thing like that and rework on if we face performances issues.

File: lib/src/Uncertainty/Distribution/DistFunc.cxx
Comment (sofianehaddad) [reply lib/src/Uncertainty/Distribution/DistFunc.cxx:1968]: yes sure

**Review by sofianehaddad — APPROVED** [review]

**Comment by regislebrun** [comment]
The main bottleneck in the getRealization() method is the call to Sample::select, as shown in the following flame graph:
perf_3f8d44f884c693f32e1519ea9e9c5e6663099fcc

with the following modification:
```
/* Get one realization of the distribution */
Point UniformOverMesh::getRealization() const
{
  const UnsignedInteger index = simplexSelection_.getRealization()[0];
  const Indices simplex(mesh_.getSimplex(index));
  const UnsignedInteger dimension = getDimension();
  Sample vertices(dimension + 1, dimension);
  for (UnsignedInteger i = 0; i <= dimension; ++i)
    {
      const Point vertex(mesh_.getVertex(simplex[i]));
      for (UnsignedInteger j = 0; j < dimension; ++j)
        vertices(i, j) = vertex[j];
    }
  return DistFunc::rUniformSimplex(vertices);
}
```
we get:
perf3

We see that the overhead between getRealization() and getSample() is minimal, and that most of the time is now spent in the creation and destruction of Point, NSI_Point & co.
Sofiane's implementation is probably the good way to go, but it has to be fixed on the two following points:
 * there is a confusion between the dimension *d* of the ambient space,given by the vertices, and the number of vertices of a simplex in this ambient space, which is equal to *d+1*
 * there is no particular work on the use of select() for dimensions greater than 3

**Comment by sofianehaddad** [comment]
Indeed, let us wait the RC to try performing this.
Thanks for the nice work @regislebrun

**Pull request merged**
